Panic on n02.validators-testnet.vega.xyz (validator testnet v0.68.4)

## Issue

A panic has been observed on the validator testnet. It seems that some nodes have no ranking score when queried which should not be possible.

**Log:**

```
2023-03-13T12:27:22.542Z        INFO        tendermint        txindex/indexer_service.go:97        indexed block exents        {"height": 157990}
2023-03-13T12:27:22.703Z        ERROR        datanode.gateway.gql        graphql/resolvers.go:2332        stream closed        {"error": "rpc error: code = Canceled desc = context canceled"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x24 pc=0x214ba26]
goroutine 3250452 [running]:
code.vegaprotocol.io/vega/datanode/sqlstore.(*Node).GetNodeData(_, {_, _}, _)
        /home/runner/work/vega/vega/datanode/sqlstore/node.go:242 +0x586
code.vegaprotocol.io/vega/datanode/api.(*tradingDataServiceV2).GetNetworkData(0xc0013c03c0, {0x3b698f8, 0xc02acb24b0}, 0xc01db7d4d8?)
        /home/runner/work/vega/vega/datanode/api/trading_data_v2.go:2303 +0x1c5
code.vegaprotocol.io/vega/protos/data-node/api/v2._TradingDataService_GetNetworkData_Handler.func1({0x3b698f8, 0xc02acb24b0}, {0x2f156a0?, 0xc02acb2450})
        /home/runner/work/vega/vega/protos/data-node/api/v2/trading_data_grpc.pb.go:3382 +0x7b
code.vegaprotocol.io/vega/datanode/ratelimit.(*RateLimit).GRPCInterceptor(0xc00360c020, {0x3b698f8, 0xc02acb24b0}, {0x2f156a0, 0xc02acb2450}, 0xb7b33d?, 0xc01daa2d80)
        /home/runner/work/vega/vega/datanode/ratelimit/ratelimit.go:136 +0x586
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x3b698f8?, 0xc02acb24b0?}, {0x2f156a0?, 0xc02acb2450?})
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:1162 +0x5b
code.vegaprotocol.io/vega/datanode/api.headersInterceptor.func1({0x3b698f8, 0xc02acb24b0}, {0x2f156a0, 0xc02acb2450}, 0x2c0ac80?, 0xc01e23e240)
        /home/runner/work/vega/vega/datanode/api/server.go:328 +0x5e3
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x3b698f8?, 0xc02acb24b0?}, {0x2f156a0?, 0xc02acb2450?})
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:1165 +0x83
code.vegaprotocol.io/vega/datanode/api.remoteAddrInterceptor.func1({0x3b698f8, 0xc02acb2420}, {0x2f156a0, 0xc02acb2450}, 0xc004341000, 0xc01e23e240)
        /home/runner/work/vega/vega/datanode/api/server.go:281 +0x637
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x3b698f8?, 0xc02acb2420?}, {0x2f156a0?, 0xc02acb2450?})
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:1165 +0x83
google.golang.org/grpc.chainUnaryInterceptors.func1({0x3b698f8, 0xc02acb2420}, {0x2f156a0, 0xc02acb2450}, 0xc004341000, 0xc01daa2d80)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:1167 +0x12b
code.vegaprotocol.io/vega/protos/data-node/api/v2._TradingDataService_GetNetworkData_Handler({0x32cc4e0?, 0xc0013c03c0}, {0x3b698f8, 0xc02acb2420}, 0xc005747030, 0xc00360c160)
        /home/runner/work/vega/vega/protos/data-node/api/v2/trading_data_grpc.pb.go:3384 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0013c01e0, {0x3b805e0, 0xc0021db860}, 0xc005877680, 0xc00358a570, 0x56a1248, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:1340 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc0013c01e0, {0x3b805e0, 0xc0021db860}, 0xc005877680, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:1713 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:963 +0x28a
```

### Additional information:

- https://vegaprotocol.slack.com/archives/C9E6XA8GK/p1678711262755359

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -60,6 +60,7 @@
 - [7750](https://github.com/vegaprotocol/vega/issues/7750) - Fix not all paths cleanly close network history index store.
 - [7805](https://github.com/vegaprotocol/vega/issues/7805) - Fix re-announcing node in the same epoch kills data node.
 - [7820](https://github.com/vegaprotocol/vega/issues/7820) - Remove the check for past date in limits engine
+- [7822](https://github.com/vegaprotocol/vega/issues/7822) - Fix get last epoch query
 
 ## 0.68.0
 

--- a/datanode/sqlstore/epochs.go
+++ b/datanode/sqlstore/epochs.go
@@ -79,7 +79,7 @@ func (es *Epochs) GetCurrent(ctx context.Context) (entities.Epoch, error) {
 	query := `WITH epochs_current AS (SELECT DISTINCT ON (id) * FROM epochs ORDER BY id DESC, vega_time DESC)
 		SELECT e.id, e.start_time, e.expire_time, e.end_time, e.tx_hash, e.vega_time, bs.height first_block, be.height last_block FROM epochs_current AS e
     	LEFT JOIN blocks bs on e.start_time = bs.vega_time
-    	LEFT JOIN blocks be on e.end_time = be.vega_time FETCH FIRST ROW ONLY;`
+    	LEFT JOIN blocks be on e.end_time = be.vega_time ORDER BY id DESC, vega_time DESC FETCH FIRST ROW ONLY;`
 
 	epoch := entities.Epoch{}
 	defer metrics.StartSQLQuery("Epochs", "GetCurrent")()

--- a/datanode/sqlstore/node.go
+++ b/datanode/sqlstore/node.go
@@ -19,6 +19,7 @@ import (
 
 	"code.vegaprotocol.io/vega/datanode/entities"
 	"code.vegaprotocol.io/vega/datanode/metrics"
+	"code.vegaprotocol.io/vega/logging"
 	v2 "code.vegaprotocol.io/vega/protos/data-node/api/v2"
 	"github.com/georgysavva/scany/pgxscan"
 )
@@ -249,6 +250,10 @@ func (store *Node) GetNodeData(ctx context.Context, epochSeq uint64) (entities.N
 		entities.ValidatorNodeStatusPending:    &nodeData.PendingNodes,
 	}
 	for _, n := range nodes {
+		if n.RankingScore == nil {
+			store.log.Warn("ignoring node with empty ranking score", logging.String("id", n.ID.String()))
+			continue
+		}
 		status := n.RankingScore.Status
 		previousStatus := n.RankingScore.PreviousStatus
 		if status == entities.ValidatorNodeStatusUnspecified {

